akka-projection-core/src/main/scala/akka/projection/internal/TelemetrySpi.scala

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.internal
+
+import scala.collection.immutable
+
+import akka.actor.typed.ActorSystem
+import akka.annotation.InternalApi
+import akka.annotation.InternalStableApi
+import akka.projection.ProjectionId
+
+/**
+ * Service Provider Interface (SPI) for collecting metrics from projections.
+ *
+ * Implementations must include a single constructor with two arguments: [[ProjectionId]]
+ * and [[ActorSystem[_]].  To setup your implementation, add a setting on your `application.conf`:
+ *
+ * {{{
+ * akka.projection.telemetry.implementation-class = com.example.MyMetrics
+ * }}}
+ */
+@InternalStableApi
+trait Telemetry {
+
+  /** Invoked when a projection is stopped. The reason for stopping is unspecified, can be a
+   * graceful stop or a failure (see [[failed()]]).
+   */
+  def stopped(): Unit
+
+  /**
+   * Invoked when a projection processing an envelope fails (even after all retry attempts).  The
+   * projection may then be restarted by a supervisor.
+   *
+   * @param cause exception thrown by the errored envelope handler.
+   */
+  def failed(cause: Throwable): Unit
+
+  /**
+   * Invoked as soon as the envelope is read, deserialised and ready to be processed.
+   *
+   * @param envelope the envelope that's ready for processing.  The type [[Envelope]] will always
+   *                 represent a single item as stored in the event log.
+   * @return an externally-provided context that will propagate with the envelope until [[afterProcess()]]
+   */
+  def beforeProcess[Envelope](envelope: Envelope): AnyRef
```

Comment:
Maybe it should be ProjectionContext instead of Envelope since additional "tracing" context could be added to the ProjectionContext. We probably need to start working on the Cinnamon side and then adopt this SPI to what is best needed there.

Reply:
`beforeProcess` is the first invocation for a given envelope. There still was no chance for telemetry to access the envelope and recreate any tracing or context. the returned value (the `AnyRef`) will _be_ that info.